executor/coprocessor.go

```diff
@@ -0,0 +1,187 @@
+		return h.buildResp(err)
+	}
+
+	chk := newFirstChunk(e)
+	tps := e.base().retFieldTypes
+	for {
+		chk.Reset()
+		err = Next(ctx, e, chk)
+		if err != nil {
+			break
+		}
+		if chk.NumRows() == 0 {
+			break
+		}
+		err = h.fillUpData4SelectResponse(chk, tps)
+		if err != nil {
+			break
+		}
+	}
+	return h.buildResp(err)
+}
+
+func (h *CoprocessorDAGHandler) buildDAGExecutor(req *coprocessor.Request) (Executor, error) {
+	if req.GetTp() != kv.ReqTypeDAG {
+		return nil, errors.Errorf("unsupported request type %d", req.GetTp())
+	}
+	dagReq := new(tipb.DAGRequest)
+	err := proto.Unmarshal(req.Data, dagReq)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	h.sctx.GetSessionVars().StmtCtx.SetFlagsFromPBFlag(dagReq.Flags)
+	h.sctx.GetSessionVars().StmtCtx.TimeZone, err = timeutil.ConstructTimeZone(dagReq.TimeZoneName, int(dagReq.TimeZoneOffset))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	h.dagReq = dagReq
+	e, err := h.buildDAG(dagReq.Executors)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return e, nil
+}
+
+func (h *CoprocessorDAGHandler) buildDAG(executors []*tipb.Executor) (Executor, error) {
+	is := h.sctx.GetSessionVars().TxnCtx.InfoSchema.(infoschema.InfoSchema)
+	bp := core.NewPBPlanBuilder(h.sctx, is)
+	plan, err := bp.BuildPhysicalPlanFromPB(executors)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	b := newExecutorBuilder(h.sctx, is)
+	return b.build(plan), nil
+}
+
+func (h *CoprocessorDAGHandler) fillUpData4SelectResponse(chk *chunk.Chunk, tps []*types.FieldType) error {
+	var err error
+	switch h.dagReq.EncodeType {
+	case tipb.EncodeType_TypeDefault:
+		err = h.encodeDefault(chk, tps)
+	case tipb.EncodeType_TypeChunk:
+		err = h.encodeChunk(chk, tps)
+	default:
+		return errors.Errorf("unknown dag encode type, %v", h.dagReq.EncodeType)
+	}
+	h.selResp.EncodeType = h.dagReq.EncodeType
+	return err
+}
+
+func (h *CoprocessorDAGHandler) buildResp(err error) *coprocessor.Response {
+	resp := h.resp
+	if err != nil {
+		resp.OtherError = err.Error()
+		return resp
+	}
+	data, err := proto.Marshal(h.selResp)
+	if err != nil {
+		resp.OtherError = err.Error()
+		return resp
+	}
+	resp.Data = data
+	return resp
+}
+
+func (h *CoprocessorDAGHandler) encodeChunk(chk *chunk.Chunk, colTypes []*types.FieldType) error {
+	colOrdinal := h.dagReq.OutputOffsets
+	chunks := h.selResp.Chunks
+	respColTypes := make([]*types.FieldType, 0, len(colOrdinal))
+	for _, ordinal := range colOrdinal {
+		respColTypes = append(respColTypes, colTypes[ordinal])
+	}
+	encoder := chunk.NewCodec(respColTypes)
+	chunks = append(chunks, tipb.Chunk{})
+	cur := &chunks[len(chunks)-1]
+	cur.RowsData = append(cur.RowsData, encoder.Encode(chk)...)
+	h.selResp.Chunks = chunks
+	return nil
+}
+
+func (h *CoprocessorDAGHandler) encodeDefault(chk *chunk.Chunk, tps []*types.FieldType) error {
+	colOrdinal := h.dagReq.OutputOffsets
+	chunks := h.selResp.Chunks
+	for i := 0; i < chk.NumRows(); i++ {
+		requestedRow := make([]byte, 0)
```

Comment:
We do not need to make a new slice for every loop.

Reply:
great, done.